cairo/ethereum/cancun/vm.cairo

```diff
@@ -88,6 +105,311 @@ struct MessageStruct {
+
+    // TODO: Move this to precompiled_contracts.cairo
+    const RIPEMD160_ADDRESS = 0x0300000000000000000000000000000000000000;
+
+    // Check if RIPEMD160 address is in child's touched accounts
+    let child_touched_accounts = child_evm.value.touched_accounts;
+    let child_touched_accounts_start = child_touched_accounts.value.dict_ptr_start;
+    let child_touched_accounts_end = cast(child_touched_accounts.value.dict_ptr, DictAccess*);
+    let (ripemd_touched) = hashdict_read{dict_ptr=child_touched_accounts_end}(
+        1, new RIPEMD160_ADDRESS
+    );
+
+    // Soundness requirement: squash all child dicts - including ones from message and env.
+
+    // EVM //
+    dict_squash(
+        cast(child_touched_accounts_start, DictAccess*),
+        cast(child_touched_accounts_end, DictAccess*),
+    );
+
+    let child_accessed_addresses = child_evm.value.accessed_addresses;
+    let child_accessed_addresses_start = child_accessed_addresses.value.dict_ptr_start;
+    let child_accessed_addresses_end = cast(child_accessed_addresses.value.dict_ptr, DictAccess*);
+    dict_squash(cast(child_accessed_addresses_start, DictAccess*), child_accessed_addresses_end);
+
+    let child_accessed_storage_keys = child_evm.value.accessed_storage_keys;
+    let child_accessed_storage_keys_start = child_accessed_storage_keys.value.dict_ptr_start;
+    let child_accessed_storage_keys_end = cast(
+        child_accessed_storage_keys.value.dict_ptr, DictAccess*
+    );
+    dict_squash(
+        cast(child_accessed_storage_keys_start, DictAccess*),
+        cast(child_accessed_storage_keys_end, DictAccess*),
+    );
+
+    let accounts_to_delete = child_evm.value.accounts_to_delete;
+    let accounts_to_delete_start = accounts_to_delete.value.dict_ptr_start;
+    let accounts_to_delete_end = cast(accounts_to_delete.value.dict_ptr, DictAccess*);
+    dict_squash(cast(accounts_to_delete_start, DictAccess*), accounts_to_delete_end);
+
+    let valid_jump_destinations = child_evm.value.valid_jump_destinations;
+    let valid_jump_destinations_start = valid_jump_destinations.value.dict_ptr_start;
+    let valid_jump_destinations_end = cast(valid_jump_destinations.value.dict_ptr, DictAccess*);
+    dict_squash(cast(valid_jump_destinations_start, DictAccess*), valid_jump_destinations_end);
+
+    let stack = child_evm.value.stack;
+    let stack_start = stack.value.dict_ptr_start;
+    let stack_end = cast(stack.value.dict_ptr, DictAccess*);
+    dict_squash(cast(stack_start, DictAccess*), cast(stack_end, DictAccess*));
+
+    let memory = child_evm.value.memory;
+    let memory_start = memory.value.dict_ptr_start;
+    let memory_end = cast(memory.value.dict_ptr, DictAccess*);
+    dict_squash(cast(memory_start, DictAccess*), cast(memory_end, DictAccess*));
+
+    // No need to squash the message's `accessed_addresses` and `accessed_storage_keys` because
+    // they were moved into the Evm, which we just squashed.
```

Comment:
I am not sure about this formulation. It should explicitly say that the those dict are shared between the message and the Evm for more context?